### PR TITLE
Implement native PyTorch support for sample_discrete_maps function

### PR DESCRIPTION
**Pull request by Akhilesh-max**
- [X] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [X] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [X] Check the commit's or even all commits' message styles matches our requested structure.

### List of changes to the codebase in this pull request:
- Added native PyTorch tensor support in sample_discrete_maps function
- Removed unnecessary conversion from PyTorch tensors to NumPy arrays
- Implemented backend-specific logic to handle both NumPy and PyTorch
- Used PyTorch's native functions like torch.unique and torch.multinomial
- Preserved device information for PyTorch tensors
- Maintained weight normalization for different backends
- Fixed TODO in mathext.py file

### CODE REVIEW

File: pgmpy/utils/mathext.py [hunk 1]
```diff
@@ -179,22 +180,58 @@ def sample_discrete_maps(states, weight_indices, index_to_weight, size=1, seed=N
...
 
-    # TODO: Remove this conversion and find a way to do this natively in torch.
-    states = np.array(states)
-    weight_indices = compat_fns.to_numpy(weight_indices)
-    index_to_weight = {
-        key: compat_fns.to_numpy(value) for key, value in index_to_weight.items()
-    }
-    size = int(size)
+    # Handle PyTorch tensors natively if we're using the torch backend
+    if config.get_backend() == "torch" and isinstance(weight_indices, torch.Tensor):
```
Comment (ankurankan) [inline pgmpy/utils/mathext.py]: @Akhilesh-max The idea is not to do this if-else block for torch and numpy. Ideally, we would like to handle both in a single block without such if-else logic. To make some of the functions agnostic to the array type we implement functions in `compat_fns.py` that applies the relevant operation depending on whether the array is numpy or torch.

File: pgmpy/utils/compat_fns.py [hunk 2]
```diff
@@ -165,3 +165,55 @@ def allclose(arr1, arr2, atol):
...
+    if config.get_backend() == "numpy":
+        return np.zeros(size, dtype=dtype or config.get_dtype())
+    else:
+        return torch.zeros(
+            size, dtype=dtype or config.get_dtype(), device=config.get_device()
+        )
+
+
+def random_choice(values, size=1, p=None, seed=None):
+    if seed is not None:
```
Comment (ankurankan) [inline pgmpy/utils/compat_fns.py:180]: Please use `np.random.default_rng` and `torch.random.fork_rng` to set seeds only for this method. Directly calling `np.random.seed` or `manual_seed` would set a global seed for them.

File: pgmpy/utils/compat_fns.py [hunk 3]
```diff
@@ -165,3 +165,55 @@ def allclose(arr1, arr2, atol):
...
+        np.random.seed(seed)
+        if torch.is_tensor(p) or torch.is_tensor(values):
+            torch.manual_seed(seed)
+
+    if isinstance(values, np.ndarray) or (
```
Comment (ankurankan) [inline pgmpy/utils/compat_fns.py:185]: I don't think we will need this many if-elses. If the backend is numpy, both `values` and `p` arrays should be numpy arrays and torch tensors otherwise. If that is not the case, we should instead try to fix the code when these `values` and `p` arrays are created instead of doing a manual check and conversion here.

**Review by ankurankan — CHANGES_REQUESTED** [review]
@Akhilesh-max I have added a couple of comments. The tests also seem to be failing. Could you have a look at them as well?

File: pgmpy/utils/compat_fns.py (hunk 3 above)
Comment (Akhilesh-max) [reply pgmpy/utils/compat_fns.py:185]: Both `values` and 'p' are not being set according to the specified backend when they are created. I am writing a fix.
